src/tmate.py

```diff
@@ -0,0 +1,198 @@
+
+
+class IncompleteInitError(Exception):
+    """The tmate-ssh-server has not been fully initialized."""
+
+
+class FingerprintError(Exception):
+    """Represents an error with generating fingerprints from public keys."""
+
+
+def install_dependencies() -> None:
+    """Install dependenciese required to start tmate-ssh-server container.
+
+    Raises:
+        DependencyInstallError: if there was something wrong installing the apt package
+            dependencies.
+    """
+    try:
+        apt.update()
+        apt.add_package(["docker.io", "openssh-client"])
+    except (apt.PackageNotFoundError, apt.PackageError) as exc:
+        raise DependencyInstallError from exc
+
+
+def install_keys(host_ip: typing.Union[ipaddress.IPv4Address, ipaddress.IPv6Address, str]) -> None:
+    """Install key creation script and generate keys.
+
+    Args:
+        host_ip: The charm host's public IP address.
+
+    Raises:
+        KeyInstallError: if there was an error creating ssh keys.
+    """
+    environment = jinja2.Environment(loader=jinja2.FileSystemLoader("templates"), autoescape=True)
+    template = environment.get_template("create_keys.sh.j2")
+    script = template.render(keys_dir=KEYS_DIR, host=str(host_ip), port=PORT)
+    WORK_DIR.mkdir(parents=True, exist_ok=True)
+    KEYS_DIR.mkdir(parents=True, exist_ok=True)
+    CREATE_KEYS_SCRIPT_PATH.write_text(script, encoding="utf-8")
+    try:
+        # B603:subprocess_without_shell_equals_true false positive
+        # see https://github.com/PyCQA/bandit/issues/333
+        subprocess.check_call(["/usr/bin/chown", "-R", "ubuntu:ubuntu", str(WORK_DIR)])  # nosec
+        CREATE_KEYS_SCRIPT_PATH.chmod(755)
+        subprocess.check_call(["/usr/bin/sudo", str(CREATE_KEYS_SCRIPT_PATH)])  # nosec
+    except subprocess.CalledProcessError as exc:
+        raise KeyInstallError from exc
+
+
+def start_daemon() -> None:
+    """Install unit files and start daemon.
+
+    Raises:
+        DaemonStartError: if there was an error starting the tmate-ssh-server docker process.
+    """
+    environment = jinja2.Environment(loader=jinja2.FileSystemLoader("templates"), autoescape=True)
+    service_content = environment.get_template("tmate-ssh-server.service.j2").render(
+        WORKDIR=WORK_DIR,
+        KEYS_DIR=KEYS_DIR,
+        PORT=PORT,
+    )
+    TMATE_SSH_SERVER_SERVICE_PATH.write_text(service_content, encoding="utf-8")
+    try:
+        # B603:subprocess_without_shell_equals_true false positive
+        # see https://github.com/PyCQA/bandit/issues/333
+        subprocess.check_call(["/usr/bin/systemctl", "daemon-reload"])  # nosec
+        subprocess.check_call(["/usr/bin/systemctl", "restart", "tmate-ssh-server"])  # nosec
+        subprocess.check_call(["/usr/bin/systemctl", "enable", "tmate-ssh-server"])  # nosec
+    except subprocess.CalledProcessError as exc:
+        raise DaemonStartError from exc
+
+
+@dataclasses.dataclass
+class Fingerprints:
```

Comment:
From [previous PR](https://github.com/canonical/tmate-ssh-server-operator/pull/2)
[@jdkandersson] Is Fingerprint one word?